cirq-core/cirq/ops/controlled_gate.py

```diff
@@ -82,29 +82,35 @@ def __init__(
         _validate_sub_object(sub_gate)
         if num_controls is None:
             if control_values is not None:
-                num_controls = len(control_values)
+                num_controls = (
+                    control_values._num_qubits_()
+                    if isinstance(control_values, cv.AbstractControlValues)
+                    else len(control_values)
+                )
             elif control_qid_shape is not None:
                 num_controls = len(control_qid_shape)
             else:
                 num_controls = 1
         if control_values is None:
             control_values = ((1,),) * num_controls
-        if num_controls != len(control_values):
-            raise ValueError('len(control_values) != num_controls')
 
-        if control_qid_shape is None:
-            control_qid_shape = (2,) * num_controls
-        if num_controls != len(control_qid_shape):
-            raise ValueError('len(control_qid_shape) != num_controls')
-        self._control_qid_shape = tuple(control_qid_shape)
-
-        # Convert to sorted tuples
+        # Convert to `cv.ProductOfSums` if input is a tuple of control values for each qubit.
```

Comment:
If there's any concern about maintaining the new classes, we could make them private and only use them as internal implementation details. We already do this here with ProductOfSums, and we could do so with SumOfProducts by allowing the control_values param to be a `Set[Sequence[int]]`, interpreting that as a ProductOfSums (which is the only natural interpretation) and doing that conversion here too. That way we allow SumOfProducts functionality, but wouldn't have to expose AbstractControlValues in the constructor parameters here, and it would allow us to continue to evolve things after 1.0.

Reply:
the reason we created the abstract class is to use it here and elsewhere wherever control values are used, the abstract functions defined in the abstract class should be the main way for interacting with control values (except in special places, like defining indexing for ProductOfSums)


this way in our code we don't have to do things like 'if isinstance' or to compute on the fly values

as for users they should pass an instance of child class of the abstract class (for now ProductOfSums or SumOfProducts) and everything will work fine, the old way of defining control values is strill defined for backward compatibility but under the hood it transforms it to a ProductOfSums

Reply:
the reason we created the abstract class is to use it here and elsewhere wherever control values are used, the abstract functions defined in the abstract class should be the main way for interacting with control values (except in special places, like defining indexing for ProductOfSums)

this way in our code we don't have to do things like 'if isinstance' or to compute on the fly values

as for users they should pass an instance of child class of the abstract class (for now ProductOfSums or SumOfProducts) and everything will work fine, the old way of defining control values is strill defined for backward compatibility but under the hood it transforms it to a ProductOfSums